src/omv/imlib/png.c

```diff
@@ -52,18 +117,196 @@ bool png_compress(image_t *src, image_t *dst, bool realloc)
+                  s += iDelta;
+              }
+              *d++ = uc; // store last partial byte
+          } else if (pDraw->iPixelType == PNG_PIXEL_GRAYSCALE)
+          {
+          uint8_t *s = pDraw->pPixels;
+          uint8_t uc, ucMask, *d = (uint8_t *)dst->data;
+          const int iPitch = IMAGE_BINARY_LINE_LEN_BYTES(dst);
+
+              d += pDraw->y * iPitch; // starting offset
+              uc = 0;
+              ucMask = 0x1;
+              for (int i=0; i<pDraw->iWidth; i++) {
+                  if (s[i] >= 128) // white
+                      uc |= ucMask;
+                  ucMask <<= 1;
+                  if (ucMask == 0) { // new byte
+                      *d++ = uc;
+                      uc = 0;
+                      ucMask = 0x1;
+                  }
+              }
+              *d++ = uc; // store last partial byte
+          }
+          break;
+      case PIXFORMAT_RGB565:
+          {
+          uint16_t *d = (uint16_t *)dst->data;
+              d += (pDraw->y * pDraw->iWidth);
+              PNGRGB565(pDraw, d, PNG_RGB565_LITTLE_ENDIAN, 0, 0);
+          }
+          break;
+// Future
+//      case PIXFORMAT_ARGB8888:
+//          break;
+   } /* switch on pixel format */
+} /* png_callback() */
 
 void png_decompress(image_t *dst, image_t *src)
 {
+PNG_DEC_IMAGE *pPNG;
+int rc; //iSize;
+
+    printf("In png_decompress\n");
+    pPNG = (PNG_DEC_IMAGE *)fb_alloc(sizeof(PNG_DEC_IMAGE), FB_ALLOC_PREFER_SIZE | FB_ALLOC_CACHE_ALIGN);
+    if (pPNG == NULL)
+        return; // DEBUG - out of memory
+    printf("allocated PNG_DEC_IMAGE structure\n");
+    memset(pPNG, 0, sizeof(PNG_DEC_IMAGE));
+    rc = PNGDEC_openRAM(pPNG, src->data, src->size, NULL);
+    if (rc != PNG_SUCCESS) {
+       fb_free(); // DEBUG - throw error here
+       printf("PNGDEC_openRAM failed!\n");
+       return;
+    }
+    printf("PNGDEC_openRAM succeeded!\n");
+    //iSize = PNGDEC_getBufferSize(pPNG);
+    //dst->data = (uint8_t *)fb_alloc(iSize, FB_ALLOC_PREFER_SIZE | FB_ALLOC_CACHE_ALIGN);
+    if (dst->data == NULL) {
+       fb_free(); // free PNG structure
+       return; // DEBUG - throw error here
+    }
+    pPNG->pfnDraw = png_callback; // give the PNG decoder a pointer to the callback function
+    rc = DecodePNG(pPNG, (void *)dst, (dst->pixfmt == PIXFORMAT_RGB565)? PNG_FAST_PALETTE : 0);
+    fb_free(); // free the PNG structure
+    if (rc != PNG_SUCCESS) {
+       printf("DecodePNG failed!\n");
+       return; // DEBUG - throw error here
+    } else {
+       printf("DecodePNG succeeded!\n");
+    }
+
     // Handle like jpeg decompress... but, one method for BINARY, GRAYSCALE, RGB565.
+    return; // good decode
 }
 
 #if defined(IMLIB_ENABLE_IMAGE_FILE_IO)
 // This function inits the geometry values of an image.
 void png_read_geometry(FIL *fp, image_t *img, const char *path, png_read_settings_t *rs)
 {
-    // Larry, see the jpeg version for how to populate the png read settings
-}
+    for (;;) {
```

Comment:
Why is the for(;;) needed.

Reply:
It was a carry over from the JPEG version; fixed now.